torch/_dynamo/eval_frame.py

```diff
@@ -262,19 +264,125 @@ def innermost_fn(fn):
+    "skipfiles_inline_module_allowlist",  # cannot pickle module (unused)
+    "constant_functions",  # PyCapsules (i.e. function ptrs)
+    "traceable_tensor_subclasses",  # Cannot pickle local object
+}
+
+# The config to restore to should dynamo compile / recompile when
+# executing from the compiled function's _TorchDynamoContext
+config_cache = threading.local()
+
+
+def _maybe_init_guarded_config_cache():
+    if hasattr(config_cache, "saved_config"):
+        return
+    else:
+        for c in ["saved_config", "saved_config_hash", "cached_recompile_hash"]:
+            setattr(config_cache, c, None)
+
+
 @contextlib.contextmanager
-def enable_dynamic(enable: Optional[bool] = None, export: bool = False):
-    if enable is None:
+def restore_compiled_dynamo_config(
+    first_ctx: bool,
+    saved_config: Dict[str, Any],
+    saved_config_hash: str,
+):
+    _maybe_init_guarded_config_cache()
+    # Set exactly once from top-level compile
+    is_top_level = False
+    try:
+        if first_ctx and config_cache.saved_config is None:
+            assert config_cache.saved_config_hash is None
+            is_top_level = True
+            config_cache.saved_config = saved_config
+            config_cache.saved_config_hash = saved_config_hash
+            log.debug(
+                "Set top-level compile config hash: %s", config_cache.saved_config_hash
+            )
+        elif config.verbose:
+            log.debug("Ignoring inner dynamo compile config and hash")  # TODO: remove?
         yield
-    elif enable:
-        # Assume everything is dynamic by default
-        with config.patch(assume_static_by_default=False):
-            yield
+    finally:
+        if is_top_level:
+            config_cache.saved_config_hash = config_cache.saved_config = None
+
+
+def get_config_and_hash(dynamic=None):
+    _maybe_init_guarded_config_cache()
+    # get current value of dynamo configs
+    dynamo_config = config.get_config_copy()
+    for k in dynamo_guarded_config_ignorelist:
+        dynamo_config.pop(k)
+
+    # Set appropriate dynamo config flags
+    if dynamic is None:
+        pass
+    elif dynamic:
+        dynamo_config.update({"assume_static_by_default": False})
     else:
-        with config.patch(
-            automatic_dynamic_shapes=False, assume_static_by_default=True
-        ):
-            yield
+        dynamo_config.update(
+            {"automatic_dynamic_shapes": False, "assume_static_by_default": True}
+        )
+
+    serialized_config = dict(dynamo_config)
+
+    # Only string-serialize for the hash
+    for key in dynamo_guarded_config_string_serialization_list:
+        serialized_config.update({key: str(dynamo_config[key])})
+
+    # Make serialization more deterministic by sorting
+    dynamo_config_bytes = pickle.dumps(dict(sorted(serialized_config.items())))
+
+    sha256_hash = hashlib.sha256(dynamo_config_bytes).hexdigest()
+    return dynamo_config, sha256_hash
```

Comment:
meh, surely we can get away with a sha1 LOL

Reply:
and I don't think there's a good reason to use a hexdigest? We're only doing equality so digest() is good enough

Reply:
> meh, surely we can get away with a sha1 LOL

Lol yeah I was even thinking md5? 128 bits is probably fine

Reply:
Apparently it's faster (not that it really matters): https://stackoverflow.com/a/2723941

Reply:
yes, md5 is probably ok too

Reply:
> and I don't think there's a good reason to use a hexdigest?

Actually, for human-redability sake, I think we should store hexdigest. It only costs 32 utf-8 bytes